src/lib/net/address.h

```diff
@@ -308,7 +308,11 @@ int tor_addr_is_null(const tor_addr_t *addr);
 int tor_addr_is_loopback(const tor_addr_t *addr);
 
 int tor_addr_is_valid(const tor_addr_t *addr, int for_listening);
-int tor_addr_is_valid_ipv4n(uint32_t v4n_addr, int for_listening);
+
+/* If for_listening is true allow v4n_addr = 0 too.
+ * else allow only non-zero v4n_addr. */
+#define tor_addr_is_valid_ipv4n(v4n_addr,for_listening) \
+        (for_listening == 1 || v4n_addr != 0)
```

Comment:
Is this really simpler than the original function?

I find it harder to read multi-part expressions on a single line, than the original code.